.github/workflows/pr.yaml

```diff
@@ -54,11 +54,37 @@ jobs:
         with:
           go-version-file: 'go.mod'
 
+      - name: Check for outdated bundle
```

Comment:
To me it looks like adding complexity for pretty much zero outcome, sorry :) 
This file bundle/manifests files are autogenerated by make bundle, so this can NOT "affect the resulting manifests" and so the message looks confusing.

Reply:
> To me it looks like adding complexity for pretty much zero outcome, sorry :)

No problem - always open to discussions :)
So the whole value here is to alleviate the load on reviewers, using some automation to streamline the review process as much as we can. This could be part of our PR template checklist, but it is always better if it can be enforced automatically, which is the point here.

> I mean I do not mind of calling extra 'make bundle' if we really need

Sure, we can think of calling `make bundle` because we know that, but if someone else contributes a PR, we would need to think about checking the bundle manifests. Even us, didn't we forget sometimes to regenerate the bundle in most of the PRs where it could have been needed? 
Same point: if you update the go code using the kubebuilder marker comments and don't regenerate the CRDs manifests (I think it already happened recently when updating the RBAC permissions), this would be caught by this validation workflow.

> sending the message about autogenerated by 'make bundle' file looks confusing for me , this can NOT "affect the resulting manifests" for github workflow.

Not sure I understand this statement. You mean the message displayed by this step, like in [this run](https://github.com/janus-idp/operator/actions/runs/7878373268/job/21496491114?pr=195#step:5:49)?
```
Files changed in bundle generation.
Please make sure to regenerate the bundle with 'make bundle' and push the changes.
Make sure you unset any related env vars like VERSION or IMAGE_TAG_BASE or IMG before running this command, as they may affect the resulting manifests.
You might also need to update the CSV in '.rhdh/bundle/manifests/rhdh-operator.csv.yaml' file accordingly.
```
This is a hint to the PR author if they want to call `make bundle` by themselves locally, to pay attention to those variables. By "affect the resulting manifests", I mean that if you had exported an env var called `VERSION` or `IMAGE_TAG_BASE`, the generated CSV will likely contain those values. Please try it by yourself, by calling `export IMAGE_TAG_BASE=my.registry.example.com/janus-idp/operator; make bundle`, and you'll see the result.
But I'm open to a different wording suggestion ;)

The CI is the source of truth, so the diff seen by the CI should be the expected outcome. Initially, I was thinking of making the GitHub job just commit the diff, so that the PR author does not even need to think about it. But for now, erroring out in case of a diff will help educate contributors to think about the resulting manifests (because the CSV in `.rhdh` might also need to be changed). That's why I'm attaching the diff to review and apply it as needed.

Reply:
Enforcing the validation sounds a good idea to me and it is commonly used by others too.

Reply:
What really concerning me is that Developer (in most of the cases triggered by this proposal) may not even have an idea that he/she changed something related.
Like I mentioned: for development I use variable like IMG to test it with my repo/image not touching prod one. It is a valid case, right?
So, let's imagine I have no idea that I did something "affecting the resulting manifests" even not changing source code and this confusing me a lot. It is happened due to particular internal implementation details and I do not think everyone should be aware of it.
TBH I thought you will make GH action provide some automation to change the things silently, I even did not consider any messaging back.

Reply:
@gazarenkov Developers should be aware that local builds can generate local-specific manifests, and before they submit a PR, they should re-generate the manifests/bundle without local specific information. This is what I have been doing and it has been working well for me:
unset IMG && unset IMAGE_TAG_BASE && unset VERSION && make manifests bundle

Reply:
> What really concerning me is that Developer (in most of the cases triggered by this proposal) may not even have an idea that he/she changed something related. Like I mentioned: for development I use variable like IMG to test it with my repo/image not touching prod one. It is a valid case, right? So, let's imagine I have no idea that I did something "affecting the resulting manifests" even not changing source code and this confusing me a lot. It is happened due to particular internal implementation details and I do not think everyone should be aware of it. 

If you did not change anything affecting the resulting manifests, then the job would just pass with no action required from you. Otherwise, you would learn that there is something needed. ;)

> TBH I thought you will make GH action provide some automation to change the things silently, I even did not consider any messaging back.

I was about to do it that way, but thought about the downstream CSV under `.rhdh`, which we may need to update as well. So IMO, we'd better enforce those with clear messaging, so that people know.
It is somehow like the lint checks to me. Sometimes, linters might be able to auto-fix some things, but in CI we tend to prefer erroring out, so that developers know for next time.

For reference, I noticed quite a few operators out there have similar validation checks (generating, then checking the git diff). See:
- https://github.com/rhobs/observability-operator/blob/main/.github/workflows/pr-checks.yaml#L60-L62
- https://github.com/os-observability/tempo-operator/blob/main/Makefile#L397-L407
- https://github.com/jaegertracing/jaeger-operator/blob/main/Makefile#L121-L127
- https://github.com/grafana/grafana-operator/blob/master/.github/workflows/pr-validation.yaml#L166-L176

Reply:
>Developers should be aware that local builds can generate local-specific manifests,

Even knowing about bundle generation specific, Developers must not keep everything in the head, keeping all the concerns unseparated is not the best way to design.
Failing workflow for fully workable code just because developer did not perform 'make bundle' locally is too confusing especially taking into account that OLM is not the only way of deploying upstream operator (even to Opeshift).

>but thought about the downstream CSV under .rhdh, which we may need to update as well.

Not sure I understand, I thought .rhdh uses manually created manifest, i.e. 'make bundle' makes no diff for it. Did I miss something?

Reply:
> Even knowing about bundle generation specific, Developers must not keep everything in the head, keeping all the concerns unseparated is not the best way to design.
Failing workflow for fully workable code just because developer did not perform 'make bundle' locally is too confusing especially taking into account that OLM is not the only way of deploying upstream operator (even to Openshift).

So that's why enforcing this in CI helps the developer IMO.
Developers are free to iterate no matter how they want locally, but we need to enforce checks after a `git push`, and this includes making sure that it meets the requirements to use the operator in production, e.g.: bundle manifests are consistent with the code changes, potential scanning tools against those manifests, ...
Sure, OLM is not the only way, but it is part of what we ship at the end of the day. So we need to enforce that it can be used consistently there as well.

> Not sure I understand, I thought .rhdh uses manually created manifest, i.e. 'make bundle' makes no diff for it. Did I miss something?
Yes, it is manually updated, but one way to remind us to consider doing so is by highlighting potential changes to the upstream CSV.

Reply:
I am OK for warning, letting developer know that it seems the locally generated code does not look the same as it is generated by CI (with possible reasons as mentioned). 

But I disagree of failing workflow for that. Unlike lint (you brought as an analogy):
- developer may not directly deal and care about this aspect (see my example above)
- this has no impact on the end result since CI's "make bundle" will produce exactly the same for synced and un-synced locally generated files

Reply:
> I am OK for warning, letting developer know that it seems the locally generated code does not look the same as it is generated by CI (with possible reasons as mentioned).
> 
> But I disagree of failing workflow for that. Unlike lint (you brought as an analogy):
> 
> * developer may not directly deal and care about this aspect (see my example above)
> * this has no impact on the end result since CI's "make bundle" will produce exactly the same for synced and un-synced locally generated files

I am not sure a warning in the validate job would be noticed, unfortunately. I think it is preferable to error out by default (since wrong bundle manifests will affect what we ship in the end).
And with the Prow bot, it is possible to override a required check manually if needed (using the `/override` comment, and providing a justification).
Again, if a developer makes changes that don't affect the manifests, nothing will change for them. The validate job will just pass. But if a change is made that changes the manifests, it would be better to know and validate it IMO.

Reply:
I understand your point but it does not look friendly and make no sense for developers who will want to try it w/o OLM, which we suggest as a quick start in fact.
For ones who care about it, warning would probably give some hint (TBH, I'd see some documentation emphasising this case) but error is too much b/c it does not impact CI.

Reply:
Ok, let me check if we can display a warning instead.. We'll see how it goes with a warning.

Reply:
@gazarenkov This should be a warning now ([`011654e` (#195)](https://github.com/janus-idp/operator/pull/195/commits/011654e3399aa3d8e7b9300d5f076e1ce8731a2b)), which won't change the outcome of the "PR Validate" job. Let's see how it goes. We can always change it back to an error in the future if it does not help that much.
To make the warning more noticeable to PR authors and reviewers, I was planning to automatically add a comment to the PR, but it does not work currently on fork PRs due to permission restrictions. This can be fixed in a subsequent issue/PR. For now, PR authors/reviewers should think about checking the validation job summary in GH for any warnings.
See https://github.com/janus-idp/operator/actions/runs/7959583931?pr=195 for an example and approve if it looks fine to you.

![image](https://github.com/janus-idp/operator/assets/593208/5cc935e9-9fb2-4f94-b47b-ff3da0db5613)

Reply:
I would agree that **warning are ignored until they're escalated to errors**. Consider the warnings about deprecated versions of nodejs used in GH actions. Until the action fails, no one is going to look at the warnings and update all their action dependencies to newer versions depending on non-EOL nodejs versions. 

![image](https://github.com/janus-idp/operator/assets/227597/44af81f8-31bb-4d49-b17e-f8288612525d)

-- https://github.com/janus-idp/operator/actions/runs/7960349847

Reply:
@nickboldt 
 I may miss something but do not think impact of this PR's warning and warnings about deprecated node.js version are comparable.
 
Let's consider something closer:  https://github.com/janus-idp/operator/issues/192
Would you consider it as "warning are ignored" and "escalate it to error" ?
If so, what concrete action items would you propose?

Reply:
my point is that if you want people to notice a problem when running a GH action, it needs to FAIL, not WARN.

Warnings are ignored; failures are addressed.